components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java

```diff
@@ -0,0 +1,399 @@
+import org.wso2.carbon.identity.action.management.model.Action;
+import org.wso2.carbon.identity.action.management.model.AuthProperty;
+import org.wso2.carbon.identity.action.management.model.AuthType;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+
+/**
+ * This class is responsible for executing the action based on the action type and the event context.
+ * It is responsible for building the request payload, calling the API, processing the response and
+ * returning the status of the action execution.
+ */
+public class ActionExecutorServiceImpl implements ActionExecutorService {
+
+    private static final Log LOG = LogFactory.getLog(ActionExecutorServiceImpl.class);
+
+    private static final ActionExecutorServiceImpl INSTANCE = new ActionExecutorServiceImpl();
+    private final APIClient apiClient;
+
+    private ActionExecutorServiceImpl() {
+
+        apiClient = new APIClient();
+    }
+
+    public static ActionExecutorServiceImpl getInstance() {
+
+        return INSTANCE;
+    }
+
+    public ActionExecutionStatus execute(ActionType actionType, Map<String, Object> eventContext, String tenantDomain)
+            throws ActionExecutionException {
+
+        try {
+            List<Action> actions = getActionsByActionType(actionType, tenantDomain);
+            validateActions(actions, actionType);
+            ActionExecutionRequest actionRequest = buildActionExecutionRequest(actionType, eventContext);
+            ActionExecutionResponseProcessor actionExecutionResponseProcessor = getResponseProcessor(actionType);
+            return executeAction(actions.get(0), actionRequest, actionType, eventContext,
+                    actionExecutionResponseProcessor);
+        } catch (ActionExecutionRuntimeException e) {
+            // todo: add to diagnostics
+            LOG.error("Skip executing actions for action type: " + actionType.name() + ". Error: " + e.getMessage(), e);
+            return new ActionExecutionStatus(ActionExecutionStatus.Status.FAILURE, eventContext);
+
+        }
+    }
+
+    private List<Action> getActionsByActionType(ActionType actionType, String tenantDomain) throws
+            ActionExecutionRuntimeException {
+
+        try {
+            return ActionExecutionServiceComponentHolder.getInstance().getActionManagementService()
+                    .getActionsByActionType(Action.ActionTypes.valueOf(actionType.name()).getPathParam(), tenantDomain);
+        } catch (ActionMgtException e) {
+            throw new ActionExecutionRuntimeException("Error occurred while retrieving actions.", e);
+        }
+    }
+
+    private void validateActions(List<Action> actions, ActionType actionType) throws ActionExecutionException {
+
+        if (actions.isEmpty()) {
```

Comment:
When there aren't any actions configured for a particular action type, actions list will be an empty list. In that case, is it required to throw a runtime exception cause if there are no actions configured we should ideally go with the basic flow as per my understanding.

Reply:
And it will Print an error log in the server logs after catching the exception in the `execute()` method. Feels like we are printing an error log for a normal behaviour.

Reply:
Fixed it such that a debug logs get printed if no actions